tessellate_ipu/linalg/tile_linalg_hessenberg.py

```diff
@@ -1,26 +1,214 @@
-from .tile_linalg_qr import ipu_qr_shard_inputs as ipu_hessenberg_shard_inputs
-from .tile_linalg_qr import qr_correction_vector_p, qr_householder_row_update_p
 
 Array = Any
 
-# Heavily based on ipu_qr_iterations in tile_linalg_qr.py
+# The code here is heavily based on tile_linalg_qr.py
+
+
+def get_hessenberg_vertex_gp_filename() -> str:
+    return os.path.join(os.path.dirname(__file__), "../core", "vertex", "tile_hessenberg_vertex.cpp")
+
+
+"""Vertex computing Hessenberg correction vector.
+"""
+hessenberg_correction_vector_p = create_ipu_tile_primitive(
+    "hessenberg_correction_vector",
+    "HessenbergCorrectionVectorVertex",
+    inputs=["Rcol", "sdiag", "cidx"],
+    outputs={"v": 0, "vrescale": ShapedArray((1,), dtype=np.float32)},
+    gp_filename=get_hessenberg_vertex_gp_filename(),
+    perf_estimate=1000,
+)
+
+"""Vertex Hessenberg HouseHolder performing row inplace update: x -= scale1[0] * scale2[0] * v
+"""
+hessenberg_householder_row_update_p = create_ipu_tile_primitive(
+    "hessenberg_householder_row_update",
+    "HessenbergHouseholderRowUpdateVertex",
+    inputs=["x", "v", "scale1", "scale2", "start_idx_"],
+    outputs={"x": 0},
+    constants={
+        "worker_offsets": lambda inavals, *_: make_ipu_vector1d_worker_offsets(
+            inavals[1].size, vector_size=2, wdtype=np.uint16
+        )
+    },
+    gp_filename=get_hessenberg_vertex_gp_filename(),
+    perf_estimate=1000,
+)
+
+
+def ipu_hessenberg_shard_inputs(x: Array, xsdiag: Array) -> Tuple[TileShardedArray, TileShardedArray, TileShardedArray]:
+    """IPU QR initial sharding of input arrays across IPU tiles.
+
+    Args:
+        x: X array.
+        sdiag: X diagonal sign.
+    Returns:
+        Tile sharded Q, R, sdiag.
+    """
+    assert x.shape[0] == x.shape[1]
+    N = x.shape[0]
+    n_tiles = 1472
+
+    # Sharding R and Q
+    if N <= 736:
+        Q_tiles = list(range(N))
+        R_tiles = list(range(N, 2 * N))
+    else:
+        n_per_tile = math.ceil(N / float(n_tiles))
+        full_tiles = N % n_tiles
+        if full_tiles == 0:
+            full_tiles = n_tiles
+
+        Q_tiles = [i for i in range(full_tiles) for _ in range(n_per_tile)] + [
+            i for i in range(full_tiles, n_tiles) for _ in range(n_per_tile - 1)
+        ]
+        R_tiles = Q_tiles
+
+    # TODO: on-device construction of identity
+    Q = tile_put_sharded(np.identity(N, dtype=x.dtype), Q_tiles)
+    R = tile_put_sharded(x, R_tiles)
+    # Replicate once on all tiles. Faster then for the looping.
+    sdiag_full = tile_put_replicated(xsdiag.T, R_tiles)
+    return Q, R, sdiag_full
+
+
 # The body of the for-loop computes
 # v = Householder(R[i])         # v is chosen to annihilate the elements below the first lower diagonal
 # R = R - 2 *  v.reshape(-1, 1) @ (v.reshape(1, -1)  @ R)
 # R = R - 2 * (R @ v.reshape(-1, 1)) @ v.reshape(1, -1)  # Not present in QR algorithm
 # Q = Q - 2 * (Q @ v.reshape(-1, 1)) @ v.reshape(1, -1)
```

Comment:
Just to be consistent with the rest of the codebase, I would move this the docstring of `ipu_hessenberg_body`

Reply:
Moved